File: examples/src/main/java/ai/freeplay/example/java/ThinExample.java
package ai.freeplay.example.java;

import ai.freeplay.client.thin.Freeplay;
import ai.freeplay.client.thin.resources.prompts.ChatMessage;
import ai.freeplay.client.thin.resources.prompts.FormattedPrompt;
import ai.freeplay.client.thin.resources.recordings.CallInfo;
import ai.freeplay.client.thin.resources.recordings.RecordInfo;
import ai.freeplay.client.thin.resources.recordings.RecordResponse;
import ai.freeplay.client.thin.resources.recordings.ResponseInfo;
import ai.freeplay.client.thin.resources.sessions.SessionInfo;
import ai.freeplay.example.java.ThinExampleUtils.Tuple3;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.http.HttpResponse;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutionException;

import static ai.freeplay.client.thin.Freeplay.Config;
import static ai.freeplay.example.java.ThinExampleUtils.callAnthropic;

public class ThinExample {
    private static final ObjectMapper objectMapper = new ObjectMapper();

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        String freeplayApiKey = System.getenv("FREEPLAY_API_KEY");
        String projectId = System.getenv("FREEPLAY_PROJECT_ID");
        String customerDomain = System.getenv("FREEPLAY_CUSTOMER_NAME");
        String anthropicApiKey = System.getenv("ANTHROPIC_API_KEY");

        Freeplay fpClient = new Freeplay(Config()
                .freeplayAPIKey(freeplayApiKey)
                .customerDomain(customerDomain)
        );

        Map<String, Object> variables = Map.of("question", "Why isn't my window working?");

        fpClient.prompts()
                .<String>getFormatted(
                        projectId,
                        "my-prompt-anthropic",
                        "prod",
                        variables,
                        null
                ).thenCompose((FormattedPrompt<String> formattedPrompt) -> {
                            long startTime = System.currentTimeMillis();
                            return callAnthropic(
                                    objectMapper,
                                    anthropicApiKey,
                                    formattedPrompt.getPromptInfo().getModel(),
                                    formattedPrompt.getPromptInfo().getModelParameters(),
                                    formattedPrompt.getFormattedPrompt()
                            ).thenApply((HttpResponse<String> response) ->
                                    new Tuple3<>(formattedPrompt, response, startTime)
                            );
                        }
                ).thenCompose((Tuple3<FormattedPrompt<String>, HttpResponse<String>, Long> promptAndResponse) -> {
                            FormattedPrompt<String> formattedPrompt = promptAndResponse.first;
                            HttpResponse<String> response = promptAndResponse.second;
                            long startTime = promptAndResponse.third;

                            JsonNode bodyNode;
                            try {
                                bodyNode = objectMapper.readTree(response.body());
                            } catch (JsonProcessingException e) {
                                throw new RuntimeException("Unable to parse response body.", e);
                            }

                            List<ChatMessage> allMessages = formattedPrompt.allMessages(
                                    new ChatMessage("Assistant", bodyNode.path("completion").asText())
                            );

                            CallInfo callInfo = CallInfo.from(
                                    formattedPrompt.getPromptInfo(),
                                    startTime,
                                    System.currentTimeMillis()
                            );
                            ResponseInfo responseInfo = new ResponseInfo(
                                    "stop_sequence".equals(bodyNode.path("stop_reason").asText())
                            );
                            SessionInfo sessionInfo = fpClient.sessions().create()
                                    .customMetadata(Map.of("custom_field", "custom_value"))
                                    .getSessionInfo();

                            System.out.println("Completion: " + bodyNode.path("completion").asText());

                            return fpClient.recordings().create(
                                    new RecordInfo(
                                            allMessages,
                                            variables,
                                            sessionInfo,
                                            formattedPrompt.getPromptInfo(),
                                            callInfo,
                                            responseInfo
                                    ));
                        }
                )
                .exceptionally(exception -> {
                    System.out.println("Got exception: " + exception.getMessage());
                    return new RecordResponse(null);
                })
                .join();
    }
}

File: examples/src/main/kotlin/ai/freeplay/example/kotlin/ThinExample.kt
package ai.freeplay.example.kotlin

import ai.freeplay.client.thin.Freeplay
import ai.freeplay.client.thin.resources.prompts.ChatMessage
import ai.freeplay.client.thin.resources.prompts.Prompts
import ai.freeplay.client.thin.resources.recordings.CallInfo
import ai.freeplay.client.thin.resources.recordings.RecordInfo
import ai.freeplay.client.thin.resources.recordings.ResponseInfo
import ai.freeplay.example.java.ThinExampleUtils.callAnthropic
import com.fasterxml.jackson.databind.ObjectMapper
import kotlinx.coroutines.future.await
import kotlinx.coroutines.runBlocking

private val objectMapper = ObjectMapper()

fun main(): Unit = runBlocking {
    val freeplayApiKey = System.getenv("FREEPLAY_API_KEY")
    val projectId = System.getenv("FREEPLAY_PROJECT_ID")
    val baseUrl = System.getenv("FREEPLAY_API_URL") + "/api"
    val anthropicApiKey = System.getenv("ANTHROPIC_API_KEY")

    val fpClient = Freeplay(
        Freeplay.Config()
            .freeplayAPIKey(freeplayApiKey)
            .baseUrl(baseUrl)
    )

    val variables = mapOf("question" to "Why isn't my window working?")

    println("Getting the prompt...")
    val prompt = fpClient
        .prompts()
        .getFormatted<List<ChatMessage>>(
            Prompts.GetFormattedRequest(
                projectId,
                "my-anthropic-prompt",
                "latest",
                variables
            )
        )
        .await()

    println("Calling Anthropic...")
    val startTime = System.currentTimeMillis()
    val llmResponse = callAnthropic(
        objectMapper,
        anthropicApiKey,
        prompt.promptInfo.model,
        prompt.promptInfo.modelParameters,
        prompt.formattedPrompt,
        prompt.systemContent.orElse(null)
    ).await()

    val endTime = System.currentTimeMillis()

    val bodyNode = objectMapper.readTree(llmResponse.body())
    println("Completion: " + bodyNode.path("content").get(0).path("text").asText())

    println("Recording the result")
    val allMessages: List<ChatMessage> = prompt.allMessages(
        ChatMessage("assistant", bodyNode.path("content").get(0).path("text").asText())
    )
    val callInfo = CallInfo.from(
        prompt.promptInfo,
        startTime,
        endTime
    )
    val responseInfo = ResponseInfo("stop_sequence" == bodyNode.path("stop_reason").asText())
    val session = fpClient.sessions().create()
        .customMetadata(mapOf("custom_field" to "custom_value"))

    val trace = session.createTrace("input str", "agent name", null)

    val recordResponse = fpClient.recordings().create(
        RecordInfo(
            projectId,
            allMessages
        ).inputs(variables)
            .sessionInfo(session.sessionInfo)
            .promptVersionInfo(prompt.promptInfo)
            .callInfo(callInfo)
            .responseInfo(responseInfo)
            .traceInfo(trace)
    ).await()

    trace.recordOutput(projectId, "output str").await()
    println("Recorded with completionId ${recordResponse.completionId}")
}
